src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardNativeWrapper.cs

```
@@ -0,0 +1,276 @@
+            buffer.Size = new UIntPtr(size);
+            buffer.Position = UIntPtr.Zero;
+        }
+
+        private void InitializeIfNotAlreadyInitialized()
+        {
+            if (!_operationInitialized)
+            {
+                _operationInitialized = true;
+
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        ZstandardNativeMethods.ZSTD_initCStream(_zstreamPointer, _compressionLevel); // start a new compression operation
+                        break;
+                    case CompressionMode.Decompress:
+                        ZstandardNativeMethods.ZSTD_initDStream(_zstreamPointer); // start a new decompression operation
+                        break;
+                }
+            }
+        }
+
+        // nested types
+        internal class OperationContext : IDisposable
+        {
+            private readonly PinnedBuffer _uncompressedPinnedBuffer;
+            private readonly PinnedBuffer _compressedPinnedBuffer;
+
+            public OperationContext(PinnedBuffer uncompressedPinnedBuffer, PinnedBuffer compressedPinnedBuffer)
+            {
+                _uncompressedPinnedBuffer = uncompressedPinnedBuffer; // can be null
+                _compressedPinnedBuffer = Ensure.IsNotNull(compressedPinnedBuffer, nameof(compressedPinnedBuffer));
+            }
+
+            public PinnedBuffer UncompressedPinnedBuffer => _uncompressedPinnedBuffer; // external data
+            public PinnedBuffer CompressedPinnedBuffer => _compressedPinnedBuffer; // internal data
+
+            // public methods
+            public void Dispose()
+            {
+                try
+                {
+                    _uncompressedPinnedBuffer?.Dispose();
+                }
+                finally
+                {
+                    _compressedPinnedBuffer.Dispose(); // ensure that both dispose methods were called
+                }
```

This use of finally to execute an unrelated operation is odd... How about:

    try { _uncompressedPinnedBuffer?.Dispose(); } catch { }
    _compressedPinnedBuffer.Dispose();

or even:

    _uncompressedPinnedBuffer?.Dispose();
    _compressedPinnedBuffer.Dispose();

Since Dispose methods are not supposed to ever throw an exception.

---

Done.
NOTE1: In general It looks like it's worth to add try/finally logic here, to be 100% sure that no unmanaged resources were lost. But since we know for sure that PinnedBuffer.Dispose suppress all errors, I'm ok to remove it.
NOTE2: I liked my original solution, but at the same time I agree that it was a little odd